### ISSUE TITLE
[Bug] EAGLE spec-decode: num_token_non_padded is 0 for draft batches, masking all draft topk ids to -1 and corrupting the MoE dispatch

### ISSUE THREAD

**Issue by Leoyzen** [sgl-project/sglang#40156]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

Since #39574 (`0dabef3d30`, "thread `num_token_non_padded` through the MoE path"), EAGLE speculative decoding with expert parallelism can hard-crash or silently corrupt output.

The EAGLE **draft** batch reaches `ForwardBatch.init_new` with `input_ids is None`, so the new count is derived as `0` instead of the real forward width. It is `0`, not `None`, so the padded-region mask is armed with a live count of zero — and that masks **every** draft topk id to `-1`. `StandardDispatcher.dispatch` then translates that `-1` through `local_expert_mapping` *before* the drop check, and PyTorch's negative-index wrap turns it into the table's last entry, which is a real local expert on the last EP rank. The MoE then dispatches a whole batch of rows into a single expert, overrunning its `m_max` region and writing out of bounds.

Chain, with line references (all on main):

1. `ForwardBatch.init_new`: `num_tokens = len(batch.input_ids) if batch.input_ids is not None else 0` → `0` for the draft batch (the scheduler rebuilds `input_ids` after `init_new`). `global_num_token_non_padded` becomes a live `0`.
2. `TopK.forward_cuda` → `select_experts` runs the padded-region mask with live count 0 → every row of the draft `topk_ids` becomes `-1`.
3. `StandardDispatcher.dispatch`: `topk_ids_local = local_expert_mapping[topk_output.topk_ids]` runs **before** the `-1` drop check. `local_expert_mapping = torch.full((288,), -1)` with `[rank*72:(rank+1)*72] = arange(72)`, so `mapping[-1] == mapping[287] == 71` on the last EP rank (benign `-1` on ranks 0–2).
4. `masked_m[71] = 48*8 = 384 > m_max = 256`; `dst = 71*256 + [0..383] = [18176, 18559]`, past the `72*256 = 18432` region → OOB write.

Observed on the last EP rank: `masked_m[71]=384`, `src2dst=[18176,18559]`, cap `18432`.

Symptoms:
- Eager: `cudaErrorIllegalAddress`. The coredump names `fill_gateup_input_triton_kernel` (OOB write); once that write is bounded, the same corruption surfaces one stage later in `silu_mul_quant_varlen_kernel`.
- CUDA graph: no crash — silent corruption. Some in-flight requests collapse to `spec_accept_rate=0.00` and emit garbage (`!!!!!…`) up to the token cap. Requests admitted during a collapse are healthy, so the corruption is per-request persistent state.
- Disabling the overlap scheduler removes it; `CUDA_LAUNCH_BLOCKING=1` removes it.

This is a main-branch regression, not a fork artifact: the fault-path files are byte-identical to `origin/main`, and `origin/main` contains the trigger line above.

### Reproduction

EAGLE spec-decode + expert parallelism, overlap scheduler + CUDA graph, high concurrency with long generation. We hit it reliably on a DeepSeek-V2-style MoE (GLM-5.3-Flash: 288 experts, top-8, `ep_size=4` → 72 local experts/rank, EAGLE topk=1 / 3 draft steps):

```bash
python3 -m sglang.launch_server \
  --model-path <deepseek-v2-style-moe> \
  --tp 4 --ep-size 4 \
  --speculative-algorithm EAGLE --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 \
  --attention-backend dsa --moe-runner-backend deep_gemm \
  # + overlap scheduler (default) and CUDA graph
```

Then drive ~48 concurrent requests with ~8k-token generations. Eager crashes with `cudaErrorIllegalAddress`; with CUDA graph some requests collapse to `accept_rate=0.00` with `!`-flood output.

A focused repro without a full server: build a draft `ForwardBatch` with `input_ids=None` and a spec info carrying `num_tokens_per_req>0`, then check that `global_num_token_non_padded == 0` and that the resulting `topk_ids` are all `-1` (which then aliases through `local_expert_mapping`).

### Environment

From `python3 -m sglang.check_env` (image `llm/sglang:glm-dcp-0916-v3`, base `dev-cu13`):

```
Python: 3.12.3
CUDA available: True
GPU 0..7: NVIDIA H20
Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: release 13.0, V13.0.88
CUDA Driver Version: 580.178.04
PyTorch: 2.13.0+cu130
sglang: 0.0.0.dev1+g8874c51a9
sglang-kernel: 0.4.7
flashinfer_python: 0.6.18
triton: 3.7.1
transformers: 5.12.1
numpy: 2.3.5
```

Related: #39574 (merged, introduced this), #39804 (open, AMD — disables the `num_token_non_padded` threading on ROCm for the same class of failure: DSV4-Pro TP8DP8 MTP GSM8K 0.119 → 0.950; it is a platform workaround, not a fix for the EAGLE draft path).

### Suggested fixes

Upstream-grade fix (two parts), plus hardening:

1. Plumb the real token count on the EAGLE draft path — the actual trigger. In `ForwardBatch.init_new`, when `input_ids is None` for a draft/spec batch, derive the width from `spec_info.num_tokens_per_req` instead of `0`, so the padded-region mask is a no-op rather than masking every row.
2. Translate only valid ids in `StandardDispatcher.dispatch` — run the `-1` drop sentinel through unchanged instead of `local_expert_mapping[topk_ids]` (e.g. `mapping[topk_ids.clamp(min=0)]` then restore `-1` with `torch.where`), so the sentinel can never alias into a real expert regardless of the count.
3. Hardening (convert silent OOB into a loud failure): bound `fill_gateup_input_triton_kernel` and `post_reorder_deepgemm`'s gather by the `[num_local_experts, m_max, *]` region; bound `silu_mul_quant_varlen_kernel`'s row index by `num_tokens`; add an async `masked_m.max() <= T` assert on the masked-GEMM consumer side (covers the DeepEP-LL path, which has no producer-side check).

Happy to open a PR if that is useful.

**Comment by he-yufeng** [comment]
I traced the chain on current main (`05e40922b3`) — part of it still holds, part has moved.

**Link 4 (the dispatcher translation) is live on main.** `token_dispatcher/standard.py` does `self.local_expert_mapping[topk_output.topk_ids]` with no drop/negative guard beforehand. A `-1` in `topk_ids` wraps to the table's last entry — which is inside the shared-experts slice when `num_local_shared_experts > 0`, i.e. a real local expert, exactly as you described. That hole is worth closing independently of the count bug (mask/drop negative ids before the mapping lookup).

**Link 1 as quoted has drifted.** `ForwardBatch.init_new` on main now does `len(batch.input_ids)` unconditionally (`forward_batch_info.py:483`), so a `None` input_ids would raise rather than silently yield 0. And the draft CUDA-graph runners (`eagle_draft_cuda_graph_runner.py:288`) build `ForwardBatch` directly with `input_ids=None` and still carry `TODO(ch-wan): support num_token_non_padded`, i.e. the padded mask is not armed on that path at all.

Could you pin the serving config where you saw this (CUDA graph on/off for the draft, `moe_ep_size`, and whether it crashed in `StandardDispatcher.dispatch`)? If the mask was armed with a live 0 somewhere, the path that derived it is the one to fix; otherwise the remaining hazard on main is the unguarded negative translation, and I can send a small PR masking negative topk ids before the mapping lookup.

**Comment by Leoyzen** [comment]
@he-yufeng Thanks for digging into this — we hit the same failure end-to-end on GLM-5.3-Flash (DSV2-style MoE, EAGLE topk1/steps3/draft4, dsa backend, deep_gemm, overlap schedule, moe_ep_size=4, 48 concurrent × 8k gen) and confirmed the whole chain with instrumented probes, not just code reading.

To your questions:

- Draft CUDA graph was on, but capture is not where it fires. `eagle_draft_cuda_graph_runner` builds the capture ForwardBatch with `input_ids=None` and never sets `num_token_non_padded` (the `TODO(ch-wan): support num_token_non_padded` at ~L643 is still there), so the padded-region mask isn't in the captured DAG. It's the eager draft fallback path that runs `_mask_topk_ids_padded_region` with a live count of 0 and masks every row to −1.
- The crash doesn't surface in `StandardDispatcher.dispatch` — the dispatcher happily translates −1 through `local_expert_mapping` and hands poisoned rows downstream. Our OOB write showed up in `fill_gateup_input_triton_kernel`: the wrap lands on the mapping table's last entry, which on EP rank 3 is local expert 71 (the shared-experts slice), so `masked_m[71]` grew to 384 > m_max=256 and dst blew past the 18432 cap. Under graph replay the same corruption is silent, which is why it presents as accept-rate collapse / garbage output rather than a traceback. Note it's rank-dependent — only the rank holding the shared-experts slice accumulates violations.

On your link-1 drift concern: current main doesn't call `len()` unconditionally. `forward_batch_info.py:987` is `num_tokens = len(batch.input_ids) if batch.input_ids is not None else 0`, so `input_ids=None` yields `num_tokens=0`, and init_new then propagates `global_num_token_non_padded = 0` (set when `enable_num_token_non_padded()`, i.e. moe_ep_size>1). That's exactly the "mask armed with a live 0" case you describe — so yes, the deriving path is the thing to fix. I think :483 is a stale line number.

Agreed the dispatcher-side guard deserves a fix independently. One subtlety though: masking negatives pre-translation must preserve the −1 sentinel (clamp(min=0) then torch.where to restore −1, keeping the pad-row drop post-translation). A bare clamp sends padded rows into the shared-experts slice as honored tokens — silent degraded output instead of a crash.

We already have this fixed and verified on our fork (0 crashes across the repro set, no accept collapse, gsm8k exact_match 0.9704, in line with baseline):

1. `ForwardBatch.init_new`: when num_tokens==0 but spec_info is present and not idle, derive the real EAGLE draft width from spec_info instead of 0 — root fix for the live-0 mask.
2. `token_dispatcher/standard.py`: clamp + where translation preserving −1 sentinels, drop stays post-translation.
3. Kernel bounds guards (fill_gateup dst bound, masked-silu token_id bound) as defense in depth, since graph replay turns residual corruption into silently wrong output.

Happy to split these into PRs against main if you'd rather review concrete diffs than write the masking patch from scratch.
